# 01 Import data

In [5]:
import pandas as pd
import numpy as np
import os

In [6]:
# importing 'products.csv' as df_prods from 'Original Data' & 'orders_wrangled.csv' as df_ords from 'Prepared Data'
path = r'C:\Users\nanh9\10.10.2024 Instacart Basket Analysis'
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [7]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.710542e+06  1.715486e+01   
std    9.875817e+05  9.875817e+05  9.875817e+05  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  8.552715e+05  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.710542e+06  1.100000e+01   
75%    2.565812e+06  2.565812e+06  2.565812e+06  2.300000e+01   
max    3.421082e+06  3.421083e+06  3.421083e+06  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

In [8]:
df_ords.drop(columns=['Unnamed: 0'], inplace =True)

## Data consistency check
### Mixed data type

In [10]:
# Create a dataframe

df_test = pd.DataFrame()

In [11]:
# Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
# Cheack for mixed data

for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [14]:
df_test['mix'] = df_test['mix'].astype('str')

### Missing values

In [16]:
# Finding missing values with isnull() function

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

=> product_name has 16 null values!

In [18]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [19]:
# Create a new dataframe: df_nan

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [20]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [21]:
# Find the mean with df.describe()
# Find the median with df_prods.median()


In [22]:
# replace missing values with mean
# df['column with missings'].fillna(mean value, inplace=True)

# replace missing value with median
# df['column with missings'].fillna(median value, inplace=True)

In [23]:
# rows & columns
df_prods.shape

(49693, 5)

In [24]:
# Create new dataframe df_prods_clean
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [25]:
df_prods_clean.shape

(49677, 5)

In [26]:
# Drop missing values
df_prods.dropna(inplace = True)

# Drop missing values only in a specific column
df_prods.dropna(subset = ['product_name'], inplace = True)

### Duplicates

In [28]:
# Finding/ idetify duplicate rows:
df_dubs = df_prods_clean[df_prods_clean.duplicated()]
df_dubs

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

#### Addressing duplicates via df_drop_duplicates()

In [30]:
df_prods_clean.shape

(49677, 5)

In [31]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [32]:
df_prods_clean_no_dups.shape

(49672, 5)

5 duplicates removed!

### Exporting data

In [35]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_cleaned.csv'))

---

# 4.5 Assignment

Cleaning up df_ords

### 2) Investigating data

In [40]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.710542e+06  1.715486e+01        2.776219e+00   
std    9.875817e+05  9.875817e+05  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  8.552715e+05  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.710542e+06  1.100000e+01        3.000000e+00   
75%    2.565812e+06  2.565812e+06  2.300000e+01        5.000000e+00   
max    3.421083e+06  3.421083e+06  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_last_order  
count       3.421083e+06           3.214874e+06  
mean        1.345202e+01           1.111484e+01  
std         4.226088e+00           9.206737e+00  
min         0.000000e+00           0.000000e+00  
25%         1.000000e+01           4.000000e+00  
50%         1.300000e+01           7.000000e+00  
75%         1.600000e+01           1.500000e+01  
max         2.300000e+01           3.000000e+01

- checking count: 'days_since_last_order' appears to have less than the rest, so there must be missing values
- checking time values: 'orders_day_of_week' with Min=0 & Max=6, range 0-6 stands for Monday to Sunday
- checking time values: 'order_hour_of_day' with Min=0 & Max=23 looks ok
- checking limits: 'order_hour_of_day' capped by 24 hours and 'orders_day_of_week' capped by 7 weekdays, which is fine
   'order_number' is capped at 100 units and 'days_since_last_order' is capped by 30 days
- data types: 'order_id', 'user_id' should be changed from integer to string                   

In [42]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329  2539329             1                   2                  8   
1   2398795  2398795             2                   3                  7   
2    473747   473747             3                   3                 12   
3   2254736  2254736             4                   4                  7   
4    431534   431534             5                   4                 15   

   days_since_last_order  
0                    NaN  
1                   15.0  
2                   21.0  
3                   29.0  
4                   28.0

### 3) Check for mixed-type data

In [44]:
# Check for mixed-type data
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

=> no output, means there is no mixed-typed data!

In [46]:
df_ords.dtypes

order_id                   int64
user_id                    int64
order_number               int64
orders_day_of_week         int64
order_hour_of_day          int64
days_since_last_order    float64
dtype: object

### 4) Fix data

### 5) Check missing values

In [49]:
df_ords.isnull().sum()

order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

We found out that 'days_since_last_order' might have missing value. Now we confirmed our guess. 'days_since_last_order' as 206209 missing values.

### 6) Address missing values

In [52]:
# Show all missing values by creating new dataframe df_ords_nan
# new dataframe df_ords_nan seperated from original

df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]
df_ords_nan.shape

(206209, 6)

first approach: replacing null values with mean, but in this case I would prefer another approach.
Doing nothing, but identify and isolate the null values.


df_ords['days_since_last_order'].fillna(7, inplace=True)

df_ords[df_ords[order_number']==1]

In [56]:
# column contains all null values -> set up a condition: all null values have order_number '1', because order_number '1' is already 'NaN' 
# Filter 'order_number' == '1'  
df_first_orders = df_ords_nan[df_ords_nan['order_number'] == '1']

In [57]:
# Check missing values for all first orders
df_first_orders_missing = df_first_orders['days_since_last_order'].isna().all()

In [58]:
# Print the boolean result True or False:
print("All first orders have missing 'days_since_last_order':", df_first_orders_missing)

All first orders have missing 'days_since_last_order': True


Every first order doesn't have a record for 'days_since_last_order' !

In [60]:
# Doublecheck if it is the same case fir the original dataframe: add new column with boolean function
df_ords['first_order'] = df_ords['days_since_last_order'].isna()
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329  2539329             1                   2   
1         2398795  2398795             2                   3   
2          473747   473747             3                   3   
3         2254736  2254736             4                   4   
4          431534   431534             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710  2266710            10                   5   
3421079   1854736  1854736            11                   4   
3421080    626363   626363            12                   1   
3421081   2977660  2977660            13                   1   
3421082    272231   272231            14                   6   

         order_hour_of_day  days_since_last_order  first_order  
0                        8                    NaN         True  
1                        7                   15.0        False  
2                       12                   21.0        False  
3                        7                   29.0        False  
4                       15                   28.0        False  
...                    ...                    ...          ...  
3421078                 18                   29.0        False  
3421079                 10                   30.0        False  
3421080                 12                   18.0        False  
3421081                 12                    7.0        False  
3421082                 14                   30.0        False  

[3421083 rows x 7 columns]

In [61]:
# data type of 'null value' should be boolean:
df_ords['first_order'].dtype

dtype('bool')

In [62]:
#summarize null values and compare counts:
df_ords.isnull().sum()

order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
first_order                   0
dtype: int64

- null values are identified
- discuss next steps with stakeholder 

### 7. Check for duplicates

In [65]:
# Finding/ idetify duplicate rows:
df_ords_dups = df_ords[df_ords.duplicated()]
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order, first_order]
Index: []

=> no duplicates detected

### 8. Handling duplicates

=> no duplicates, so no action needed

- If there were any duplicates, we need to clarify them before removing them.
- next step: use function: df_ords_nodup.drop_duplicates()
- here: df_ords_no_dups = df_ords_no_dups.drop_duplicates()
- df_ords_no_dups with no duplicates can be checked with df_ords_no_dups.shape
- compare df_ords.shape <---> df_ords_no_dups 

In [70]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [71]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_cleaned.csv'))

In [72]:
# Check number of unique product ids
df_unique_product_ids = df_prods_clean_no_dups['product_id'].nunique()

In [73]:
print(f"Number of unique product IDs: {df_unique_product_ids}")

Number of unique product IDs: 49670


In [74]:
# Check if product_id values are sequential and without gaps
df_all_ids = set(range(1, df_prods_clean_no_dups['product_id'].max() + 1))

In [75]:
# Create a set of actual ids in the dataframe
df_actual_ids = set(df_prods_clean_no_dups['product_id'].unique())

In [76]:
# Find any missing ids
missing_ids = df_all_ids - df_actual_ids

In [77]:
if missing_ids:
    print(f"There are missing product IDs: {sorted(missing_ids)}")
else:
    print("No missing product IDs; they are sequential.")

There are missing product IDs: [34, 69, 116, 262, 525, 1511, 1780, 2240, 2586, 3159, 3230, 3736, 4283, 4790, 6799, 26519, 38183, 40440]


In [78]:
outliers = df_prods_clean_no_dups[df_prods_clean_no_dups['prices'] == 99999]

In [79]:
print(outliers)

       product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0


In [80]:
# Now that we've determined where the outlier is located, there are several approaches that can be taken
# In this case, since the price skews that data immensely and it's clear that its an error and not an expensive product, we can remove it
df_prods_extra_clean_no_dups = df_prods_clean_no_dups[df_prods_clean_no_dups['prices'] != 99999]

In [81]:
df_prods_extra_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49671.000000  49671.000000   49671.000000  49671.000000
mean   24850.172334     67.762115      11.728856      7.980256
std    14340.795118     38.316100       5.850806     66.952504
min        1.000000      1.000000       1.000000      1.000000
25%    12432.500000     35.000000       7.000000      4.100000
50%    24850.000000     69.000000      13.000000      7.100000
75%    37268.500000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  14900.000000

In [82]:
# It seems like there's yet another outlier. As such, we'll investigate the rows with the highest prices to see the other outliers
df_prods_sorted = df_prods_extra_clean_no_dups.sort_values(by='prices', ascending=False)

In [83]:
top_priced_products = df_prods_sorted.head()

In [84]:
print(top_priced_products)

       product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
9020         9020  Boneless Skinless Chicken Thighs        35             12   
40490       40486                   Chicken Tenders        49             12   
21468       21467            Wild Caught Raw Shrimp        15             12   
25580       25579     Naturally Smoked Trout Fillet        15             12   

        prices  
21554  14900.0  
9020      25.0  
40490     25.0  
21468     25.0  
25580     25.0  


In [85]:
# Now we know that the product priced 14900 is the only other outlier that we must remove
df_prods_extra_clean_no_dups_2 = df_prods_extra_clean_no_dups[df_prods_extra_clean_no_dups['prices'] != 14900]

In [86]:
df_prods_extra_clean_no_dups_2.describe()

product_id      aisle_id  department_id        prices
count  49670.000000  49670.000000   49670.000000  49670.000000
mean   24850.238716     67.761305      11.728770      7.680437
std    14340.931850     38.316060       5.850834      4.199381
min        1.000000      1.000000       1.000000      1.000000
25%    12432.250000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.750000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

In [87]:
# The descriptive statistics now appear more normal and no visible errors are present

### 9. Export data to "Prepared Data" folder

In [89]:
df_prods_extra_clean_no_dups_2.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_cleaned.csv'))
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned.csv'))